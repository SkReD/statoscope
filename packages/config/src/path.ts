import path from 'path';
import module from 'module';
import chalk from 'chalk';

export enum PackageAliasType {
  PLUGIN = 'PLUGIN',
  REPORTER = 'REPORTER',
}

export function normalizePath(source: string, rootDir: string): string {
  return source.replace('<rootDir>', rootDir);
}

export function makeRequireFromPath(pathname: string): NodeRequire {
  return module.createRequire(path.resolve(pathname, '_'));
}

// package namespace (empty string when not set) followed by the package name (wo namespace)
type PackageName = [string, string];

const packageAliasPrefixes = new Map<PackageAliasType, Record<PrefixType, string>>();

enum PrefixType {
  STATOSCOPE_NS = '@statoscope',
  NON_STATOSCOPE_NS = '@NON_STATOSCOPE_NS',
}

packageAliasPrefixes.set(PackageAliasType.PLUGIN, {
  [PrefixType.STATOSCOPE_NS]: 'stats-validator-plugin',
  [PrefixType.NON_STATOSCOPE_NS]: 'statoscope-stats-validator-plugin',
});

packageAliasPrefixes.set(PackageAliasType.REPORTER, {
  [PrefixType.STATOSCOPE_NS]: 'stats-validator-reporter',
  [PrefixType.NON_STATOSCOPE_NS]: 'statoscope-stats-validator-reporter',
});

/**
 * Concatenate namespace (when provided) with the package name.
 */
function getFullPackageName([namespace, name]: PackageName): string {
  if (namespace) {
    return `${namespace}/${name}`;
  }

  return name;
}

/**
 * Return error message based on the list of package aliases (provided by the user and alternative).
 */
function getAliasPackageResolutionError(packageAliases: PackageName[]): string {
  const providedAlias = getFullPackageName(packageAliases[0]);

  let errorMessage = `Can't resolve package ${chalk.yellow.italic(providedAlias)}.\n`;

  const { italic: italicChalk } = chalk;

  if (packageAliases.length > 1) {
    const alternativeAlias = getFullPackageName(packageAliases[1]);

    errorMessage += `Also tried to resolve it with ${italicChalk.yellow(
      alternativeAlias,
    )} alias which didn't work either.\n\n`;
  } else {
    errorMessage += '\n\n';
  }

  errorMessage += 'Try installing the package locally:\n';
  errorMessage += `- with ${italicChalk('npm')}: ${chalk.bgGrey(
    'npm i -D ' + providedAlias,
  )} (or corresponding package alias)\n`;
  errorMessage += `- with ${italicChalk('yarn')}: ${chalk.bgGrey(
    'yarn add -D ' + providedAlias,
  )} (or corresponding package alias)\n`;

  return errorMessage;
}

/**
 * Resolve full or short package name to it's absolute path.
 * @param packageAliasType - Plugin or reporter enum.
 * @param aliasName - Package name (alias). Can be passed in the "full" or "short" form.
 *   Will append prefixes based on packageAliasType value in order to resolve the short form alias.
 * @param fromDir - Directory used as "root" while resolving the package.
 */
export function resolveAliasPackage(
  packageAliasType: PackageAliasType,
  aliasName: string,
  fromDir: string,
): string {
  const localRequire = makeRequireFromPath(fromDir);
  aliasName = normalizePath(aliasName, fromDir);

  if (aliasName.startsWith('.') || path.isAbsolute(aliasName)) {
    localRequire(aliasName);
    return aliasName;
  }

  let packageNamespace = '';
  let packageName = aliasName;

  const packageNameRegex = /^(@.+?)\/(.+)/;

  if (packageNameRegex.test(aliasName)) {
    [, packageNamespace = '', packageName] = aliasName.match(packageNameRegex)!;
  }

  // can have from one up to two entries
  const packageAliases: PackageName[] = [
    [packageNamespace, packageName], // original form
  ];

  const prefixes = packageAliasPrefixes.get(packageAliasType)!;

  const prefix =
    packageNamespace === PrefixType.STATOSCOPE_NS
      ? prefixes[PrefixType.STATOSCOPE_NS]
      : prefixes[PrefixType.NON_STATOSCOPE_NS];

  if (!packageName.startsWith(prefix)) {
    packageAliases.push([packageNamespace, `${prefix}-${packageName}`]);
  }

  const paths = packageAliases.map(([namespace, packageName]) =>
    path.join(namespace, packageName),
  );

  for (const path of paths) {
    try {
      localRequire(path);
      return path;
      // eslint-disable-next-line no-empty
    } catch (e) {}
  }

  throw new Error(getAliasPackageResolutionError(packageAliases));
}
