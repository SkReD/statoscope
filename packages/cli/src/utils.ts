import path from 'path';
import os from 'os';
import fs from 'fs';
import {
  FromItem,
  makeReplacer,
  transform as transformOriginal,
} from '@statoscope/report-writer/dist/utils';
import { parseChunked } from '@discoveryjs/json-ext';
import normalizeCompilation from '@statoscope/webpack-model/dist/normalizeCompilation';
import type { Webpack } from '@statoscope/webpack-model/webpack';
import { Report } from '@statoscope/types/types/custom-report';
import { Config } from '@statoscope/types/types/config';
import { Extension } from '@statoscope/stats/spec/extension';
import Generator, {
  Payload,
} from '@statoscope/stats-extension-custom-reports/dist/generator';

export type TransformFrom = {
  name: string;
  as: string;
};

export function mergeCustomReportsIntoCompilation(
  parsed: Webpack.Compilation,
  reports: unknown[],
): Webpack.Compilation {
  parsed.__statoscope ??= {};
  parsed.__statoscope.extensions ??= [];
  const customReportsExtensionIx = parsed.__statoscope.extensions.findIndex(
    (ext) => ext.descriptor.name === '@statoscope/stats-extension-custom-reports',
  );
  const customReportsExtension =
    customReportsExtensionIx > -1
      ? parsed.__statoscope.extensions[customReportsExtensionIx]
      : null;
  const customReportGenerator = new Generator();
  if (customReportsExtension?.payload) {
    const payload = customReportsExtension.payload as Payload;

    for (const compilationItem of payload.compilations) {
      for (const report of compilationItem.reports) {
        customReportGenerator.handleReport(report);
      }
    }
  }

  for (const report of reports) {
    if (isCustomReport(report)) {
      customReportGenerator.handleReport(report);
    } else {
      throw new Error(
        `Can't add a report. A valid report should contain id and view fields`,
      );
    }
  }

  if (customReportsExtension) {
    parsed.__statoscope.extensions.splice(customReportsExtensionIx, 1);
  }

  parsed.__statoscope.extensions.push(customReportGenerator.get());

  return parsed;
}

export function mergeCustomExtensionsIntoCompilation(
  parsed: Webpack.Compilation,
  extensions: Extension<unknown>[],
): Webpack.Compilation {
  parsed.__statoscope ??= {};
  parsed.__statoscope.extensions ??= [];

  for (const ext of extensions) {
    if (
      parsed.__statoscope.extensions.some(
        (e) => e.descriptor.name === ext.descriptor.name,
      )
    ) {
      console.log(`Extension ${ext.descriptor.name} already exists. Ignoring`);
      continue;
    }

    parsed.__statoscope.extensions.push(ext);
  }

  return parsed;
}

export async function transform(
  from: Array<TransformFrom | string>,
  to: string,
  customReports: Report<unknown, unknown>[] = [],
  compression: boolean,
): Promise<string> {
  const normalizedFrom: FromItem[] = [];

  for (const item of from) {
    const filename = typeof item === 'string' ? item : item.name;
    const as = typeof item === 'string' ? item : item.as;
    let parsed: Webpack.Compilation = await parseChunked(fs.createReadStream(filename));

    if (customReports.length) {
      parsed = mergeCustomReportsIntoCompilation(parsed, customReports);
    }

    normalizeCompilation(parsed);

    normalizedFrom.push({
      type: 'data',
      filename: as,
      data: parsed,
      replacer: makeReplacer(parsed.__statoscope?.context, '.', ['context', 'source']),
    });
  }

  return transformOriginal(
    {
      writer: {
        scripts: [{ type: 'path', path: require.resolve('@statoscope/webpack-ui') }],
        init: `function (data) {
          Statoscope.default(data.map((item) => ({ name: item.id, data: item.data })));
        }`,
        dataCompression: compression,
      },
    },
    normalizedFrom,
    to,
  );
}

export const createDestStatReportPath = (from: string[], to?: string): string => {
  const id = path.basename(from[0], '.json');
  return to || path.join(os.tmpdir(), `statoscope-report-${id}-${Date.now()}.html`);
};

/**
 * Return true for objects with properties expected from custom report.
 * Type guard.
 */
export function isCustomReport(report: unknown): report is Report<unknown, unknown> {
  return !!(
    report &&
    typeof report === 'object' &&
    report.constructor === Object &&
    'id' in report &&
    'view' in report
  );
}

/**
 * Combine custom user reports from statoscope config and command arguments
 * (report json file paths).
 */
export function combineCustomReports(
  config: Config,
  reportArg: string[] = [],
): Report<unknown, unknown>[] {
  const reports: Report<unknown, unknown>[] = [];

  if (reportArg) {
    let malformedReport = false;

    const reportsFromFiles: unknown[] = reportArg.flatMap((filepath) =>
      require(path.resolve(filepath)),
    );

    for (const report of reportsFromFiles) {
      if (isCustomReport(report)) {
        reports.push(report);
      } else {
        malformedReport = true;
      }
    }

    if (malformedReport) {
      console.warn('One or more custom reports from files passed are malformed');
    }
  }

  if (config.generate?.reports) {
    reports.push(...config.generate.reports);
  }

  return reports;
}
