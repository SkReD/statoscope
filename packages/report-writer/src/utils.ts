import fs from 'fs';
import path from 'path';
import { Readable, Writable } from 'stream';
import pako from 'pako';
import { parseChunked, stringifyStream, TReplacer } from '@discoveryjs/json-ext';
import HTMLWriter, { encodeBinaryJSON, Options } from './';

export function waitFinished(stream: Readable | Writable): Promise<void> {
  return new Promise((resolve, reject) => {
    stream.once('end', resolve);
    stream.once('finish', resolve);
    stream.once('error', reject);
  });
}

export type FromItem =
  | {
      type: 'data';
      filename: string;
      data: unknown;
      replacer?: TReplacer;
    }
  | {
      type: 'filename';
      filename: string;
      replacer?: TReplacer;
    }
  | {
      type: 'stream';
      filename: string;
      stream: Readable;
    };

export async function transform(
  options: {
    writer: Options;
  },
  from: Array<string | FromItem>,
  to: string,
): Promise<string> {
  const normalizedFrom: FromItem[] = from.map((item) =>
    typeof item === 'string' ? { type: 'filename', filename: item } : item,
  );
  const toDir = path.dirname(to);
  if (!fs.existsSync(toDir)) {
    fs.mkdirSync(toDir, { recursive: true });
  }
  const outputStream = fs.createWriteStream(to);
  const htmlWriter = new HTMLWriter(options.writer);

  for (const fromItem of normalizedFrom) {
    const id = path.basename(fromItem.filename);
    let stream: Readable;

    if (fromItem.type === 'filename') {
      stream = fs.createReadStream(fromItem.filename);
    } else if (fromItem.type === 'data') {
      stream = stringifyStream(fromItem.data, fromItem.replacer);
    } else {
      stream = fromItem.stream;
    }

    if (options.writer.dataCompression === false) {
      htmlWriter.addChunkWriter(stream, id);
    } else {
      const htmlStats = await parseChunked(stream);
      const binaryJSONArray = encodeBinaryJSON(htmlStats);
      const binaryJSONCompressedArray = pako.deflate(binaryJSONArray);
      let readIndex = 0;
      const htmlStatsStream = new Readable({
        read(size): void {
          const chunk = binaryJSONCompressedArray.subarray(readIndex, readIndex + size);
          readIndex += size;
          this.push(Buffer.from(chunk).toString('base64'));
          if (readIndex >= binaryJSONCompressedArray.length) {
            this.push(null);
          }
        },
      });

      htmlWriter.addChunkWriter(htmlStatsStream, id);
    }
  }

  htmlWriter.getStream().pipe(outputStream);
  htmlWriter.write();

  await waitFinished(outputStream);

  return to;
}

export function makeReplacer(
  from?: string,
  to = '',
  ignoreKeys: string[] = [],
): TReplacer | undefined {
  if (!from) {
    return;
  }

  return (key, value): unknown => {
    if (typeof value === 'string' && !ignoreKeys.includes(key)) {
      if (value.includes(from)) return value.split(from).join(to);
    }

    return value;
  };
}
