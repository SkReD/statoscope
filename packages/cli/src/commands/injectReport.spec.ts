import fs from 'fs';
import path from 'path';
import { mergeCustomReportsIntoCompilation } from '../utils';

const rootPath = path.resolve(__dirname, '../../../../');
const outputDir = path.join(
  rootPath,
  'test/temp',
  Date.now().toString(),
  path.relative(rootPath, __filename),
);
const statsDir = path.resolve(
  __dirname,
  '../../../../test/fixtures/cli/injectReport/stats/',
);
const stats = [
  'stats-with-reports.json',
  'stats-without-extensions.json',
  'stats-without-meta.json',
  'stats-without-reports.json',
];

fs.mkdirSync(outputDir, { recursive: true });

jest.mock('../../package.json', () => ({
  ...jest.requireActual('../../package.json'),
  version: '8.8.8',
}));
jest.mock('@statoscope/stats-extension-custom-reports/package.json', () => ({
  ...jest.requireActual('@statoscope/stats-extension-custom-reports/package.json'),
  version: '8.8.8',
}));

const reports = [
  require.resolve(
    '../../../../test/fixtures/cli/injectReport/reports/single-report-a.json',
  ),
  require.resolve(
    '../../../../test/fixtures/cli/injectReport/reports/multiple-reports.json',
  ),
];

test.each(stats)('%s,', async (filename) => {
  for (const report of reports) {
    const parsed = JSON.parse(fs.readFileSync(path.join(statsDir, filename), 'utf-8'));
    const parsedReport = JSON.parse(fs.readFileSync(report, 'utf-8'));

    expect(
      await mergeCustomReportsIntoCompilation(parsed, [parsedReport].flat()),
    ).toMatchSnapshot();
  }
});
