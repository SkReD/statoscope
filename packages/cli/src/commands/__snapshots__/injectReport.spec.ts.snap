// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stats-with-reports.json, 1`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "name": "foo",
        },
      },
      {
        "descriptor": {
          "adapter": undefined,
          "author": {
            "name": "Sergey Melyukov",
            "url": "https://twitter.com/smelukov",
          },
          "description": "Statoscope extension to store custom reports in stats",
          "homepage": "https://github.com/statoscope/statoscope/",
          "name": "@statoscope/stats-extension-custom-reports",
          "version": "8.8.8",
        },
        "payload": {
          "compilations": [
            {
              "id": null,
              "reports": [
                {
                  "id": "baz",
                  "view": [],
                },
                {
                  "data": [
                    1,
                    2,
                    3,
                  ],
                  "id": "foo",
                  "view": [
                    "struct",
                  ],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`stats-with-reports.json, 2`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "name": "foo",
        },
      },
      {
        "descriptor": {
          "adapter": undefined,
          "author": {
            "name": "Sergey Melyukov",
            "url": "https://twitter.com/smelukov",
          },
          "description": "Statoscope extension to store custom reports in stats",
          "homepage": "https://github.com/statoscope/statoscope/",
          "name": "@statoscope/stats-extension-custom-reports",
          "version": "8.8.8",
        },
        "payload": {
          "compilations": [
            {
              "id": null,
              "reports": [
                {
                  "id": "baz",
                  "view": [],
                },
                {
                  "data": [
                    1,
                    2,
                    3,
                  ],
                  "id": "foo",
                  "view": [
                    "list",
                  ],
                },
                {
                  "data": [
                    4,
                    5,
                    6,
                  ],
                  "id": "bar",
                  "view": [
                    "list",
                  ],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`stats-without-extensions.json, 1`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "adapter": undefined,
          "author": {
            "name": "Sergey Melyukov",
            "url": "https://twitter.com/smelukov",
          },
          "description": "Statoscope extension to store custom reports in stats",
          "homepage": "https://github.com/statoscope/statoscope/",
          "name": "@statoscope/stats-extension-custom-reports",
          "version": "8.8.8",
        },
        "payload": {
          "compilations": [
            {
              "id": null,
              "reports": [
                {
                  "data": [
                    1,
                    2,
                    3,
                  ],
                  "id": "foo",
                  "view": [
                    "struct",
                  ],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`stats-without-extensions.json, 2`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "adapter": undefined,
          "author": {
            "name": "Sergey Melyukov",
            "url": "https://twitter.com/smelukov",
          },
          "description": "Statoscope extension to store custom reports in stats",
          "homepage": "https://github.com/statoscope/statoscope/",
          "name": "@statoscope/stats-extension-custom-reports",
          "version": "8.8.8",
        },
        "payload": {
          "compilations": [
            {
              "id": null,
              "reports": [
                {
                  "data": [
                    1,
                    2,
                    3,
                  ],
                  "id": "foo",
                  "view": [
                    "list",
                  ],
                },
                {
                  "data": [
                    4,
                    5,
                    6,
                  ],
                  "id": "bar",
                  "view": [
                    "list",
                  ],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`stats-without-meta.json, 1`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "adapter": undefined,
          "author": {
            "name": "Sergey Melyukov",
            "url": "https://twitter.com/smelukov",
          },
          "description": "Statoscope extension to store custom reports in stats",
          "homepage": "https://github.com/statoscope/statoscope/",
          "name": "@statoscope/stats-extension-custom-reports",
          "version": "8.8.8",
        },
        "payload": {
          "compilations": [
            {
              "id": null,
              "reports": [
                {
                  "data": [
                    1,
                    2,
                    3,
                  ],
                  "id": "foo",
                  "view": [
                    "struct",
                  ],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`stats-without-meta.json, 2`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "adapter": undefined,
          "author": {
            "name": "Sergey Melyukov",
            "url": "https://twitter.com/smelukov",
          },
          "description": "Statoscope extension to store custom reports in stats",
          "homepage": "https://github.com/statoscope/statoscope/",
          "name": "@statoscope/stats-extension-custom-reports",
          "version": "8.8.8",
        },
        "payload": {
          "compilations": [
            {
              "id": null,
              "reports": [
                {
                  "data": [
                    1,
                    2,
                    3,
                  ],
                  "id": "foo",
                  "view": [
                    "list",
                  ],
                },
                {
                  "data": [
                    4,
                    5,
                    6,
                  ],
                  "id": "bar",
                  "view": [
                    "list",
                  ],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`stats-without-reports.json, 1`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "name": "foo",
        },
      },
      {
        "descriptor": {
          "adapter": undefined,
          "author": {
            "name": "Sergey Melyukov",
            "url": "https://twitter.com/smelukov",
          },
          "description": "Statoscope extension to store custom reports in stats",
          "homepage": "https://github.com/statoscope/statoscope/",
          "name": "@statoscope/stats-extension-custom-reports",
          "version": "8.8.8",
        },
        "payload": {
          "compilations": [
            {
              "id": null,
              "reports": [
                {
                  "data": [
                    1,
                    2,
                    3,
                  ],
                  "id": "foo",
                  "view": [
                    "struct",
                  ],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;

exports[`stats-without-reports.json, 2`] = `
{
  "__statoscope": {
    "extensions": [
      {
        "descriptor": {
          "name": "foo",
        },
      },
      {
        "descriptor": {
          "adapter": undefined,
          "author": {
            "name": "Sergey Melyukov",
            "url": "https://twitter.com/smelukov",
          },
          "description": "Statoscope extension to store custom reports in stats",
          "homepage": "https://github.com/statoscope/statoscope/",
          "name": "@statoscope/stats-extension-custom-reports",
          "version": "8.8.8",
        },
        "payload": {
          "compilations": [
            {
              "id": null,
              "reports": [
                {
                  "data": [
                    1,
                    2,
                    3,
                  ],
                  "id": "foo",
                  "view": [
                    "list",
                  ],
                },
                {
                  "data": [
                    4,
                    5,
                    6,
                  ],
                  "id": "bar",
                  "view": [
                    "list",
                  ],
                },
              ],
            },
          ],
        },
      },
    ],
  },
}
`;
